test/test_heevd_device.cc

```diff
@@ -0,0 +1,207 @@
+{
+    using lapack::device_info_int;
+    using real_t = blas::real_type< scalar_t >;
+
+    // Constants
+    const scalar_t one  = 1.0;
+    const real_t   eps  = std::numeric_limits< real_t >::epsilon();
+
+    // get & mark input values
+    lapack::Job jobz = params.jobz();
+    lapack::Uplo uplo = params.uplo();
+    int64_t n = params.dim.n();
+    int64_t device = params.device();
+    int64_t align = params.align();
+    int64_t verbose = params.verbose();
+    real_t tol = params.tol() * eps;
+    params.matrix.mark();
+
+    // mark non-standard output values
+    params.ref_time();
+    // params.ref_gflops();
+    // params.gflops();
+    params.error2();
+
+    if (! run)
+        return;
+
+    if (blas::get_device_count() == 0) {
+        params.msg() = "skipping: no GPU devices or no GPU support";
+        return;
+    }
+
+    // ---------- setup
+    int64_t lda = roundup( blas::max( 1, n ), align );
+    int64_t ldz = lda;  // vectors overwrite matrix A
+    size_t size_A = (size_t) lda * n;
+    size_t size_Z = size_A;
+    size_t size_W = (size_t) n;
+
+    std::vector< scalar_t > A( size_A );
+    std::vector< scalar_t > Z( size_Z );  // eigenvectors
+    std::vector< real_t > Lambda_tst( n );
+    std::vector< real_t > Lambda_ref( n );
+
+    lapack::generate_matrix( params.matrix, n, n, &A[0], lda );
+    Z = A;
+
+    // Allocate and copy to GPU
+    lapack::Queue queue( device );
+    scalar_t*        dA_tst = blas::device_malloc< scalar_t >( size_A, queue );
+    real_t*          dW_tst = blas::device_malloc< real_t >  ( size_W, queue );
+    device_info_int* d_info = blas::device_malloc< device_info_int >( 1, queue );
+    blas::device_copy_matrix( n, n, A.data(), lda, dA_tst, lda, queue );
+
+
+    // Allocate workspace
+    size_t d_size, h_size;
+    lapack::heevd_work_size_bytes( jobz, uplo, n, dA_tst, lda, dW_tst, 
+      &d_size, &h_size, queue );
+    char* d_work = blas::device_malloc< char >( d_size, queue );
+    std::vector<char> h_work_vector( h_size );
+    char* h_work = h_work_vector.data();
+
+
+    if (verbose >= 1) {
+        printf( "\n" );
+        printf( "A n=%5lld, lda=%5lld\n", llong( n ), llong( lda ) );
+    }
+    if (verbose >= 2) {
+        printf( "A = " ); print_matrix( n, n, &A[0], lda );
+    }
+
+    // ---------- run test
+    testsweeper::flush_cache( params.cache() );
+    queue.sync();
+    double time = testsweeper::get_wtime();
+
+    lapack::heevd( jobz, uplo, n, dA_tst, lda, dW_tst, d_work, d_size,
+      h_work, h_size, d_info, queue );
+
+    queue.sync();
+    time = testsweeper::get_wtime() - time;
+
+    params.time() = time;
+    // double gflop = lapack::Gflop< scalar_t >::heev( jobz, n );
+    // params.gflops() = gflop / time;
+
+    // Copy result back to CPU.
+    device_info_int info_tst;
+    blas::device_copy_matrix( n, n, dA_tst, lda, Z.data(), ldz, queue );
+    blas::device_copy_matrix( n, 1, dW_tst, n,   Lambda_tst.data(), n, queue );
```

Comment:
Perhaps copy_vector would be clearer here than copying an n-by-1 matrix with leading dimension n.